src/edu/csus/ecs/pc2/clics/API202306/EventFeedService.java

```diff
@@ -0,0 +1,183 @@
+        }
+
+        if (startingToken != null) {
+            if (startingToken.startsWith(EventFeedJSON.EVENT_ID_PREFIX) && Utilities.isIntegerNumber(startingToken.substring(EventFeedJSON.EVENT_ID_PREFIX_LENGTH))) {
+                filter.addStartingEventId(startingToken);
+                startMsg += "starting after id " + startingToken;
+            } else {
+                response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Not starting event feed: Invalid starting id: `"+startingToken+"`");
+                return;
+            }
+        } else {
+            startMsg += "(whole feed)";
+        }
+        info(startMsg);
+
+        /**
+         * Add stream and write past events to stream.
+         */
+        eventFeedStreamer.addStream(servletOutputStream, filter);
+
+        if (!eventFeedStreamer.isRunning()) {
+            /**
+             * Put on thread if not running on a thread.
+             */
+            new Thread(eventFeedStreamer).start();
+        }
+
+        while (true) {
+            // If the contest has been finalized or this client is not longer connected (eg not receiving the EF), we are done
+            if (eventFeedStreamer.isFinalized() || !eventFeedStreamer.isStreamActive(servletOutputStream)) {
+                break;
+            }
+            try {
+                Thread.sleep(1 * Constants.MS_PER_SECOND);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+                log.log(Level.WARNING, "During sleep " + e.getMessage());
+                break;
+            }
+        }
+        info("Event Feeder client " + filter.getClient() + " is no longer connected.");
+    }
+
+    public void info(String message) {
+        if (log != null) {
+            log.log(Level.INFO, message);
+        }
+    }
+
+    /**
+     * Retrieve access information about this endpoint for the supplied user's security context
+     *
+     * @param sc User's security information
+     * @return CLICSEndpoint object if the user can access this endpoint's properties, null otherwise
+     */
+    public static CLICSEndpoint getEndpointProperties(SecurityContext sc) {
+        String [] efProps = { "type", "id", "data", "token" };
+        return(new CLICSEndpoint("event-feed", efProps));
+    }
+
+    @Override
+    public boolean configure(FeatureContext arg0) {
+        // TODO Auto-generated method stub
+        return false;
+    }
+
+    /**
+     * Create a snapshot of the JSON event feed.
+     *
+     * @param contest
+     * @param controller
+     * @return
+     */
+    public static String createEventFeedJSON(IInternalContest contest, IInternalController controller, HttpServletRequest servletRequest, SecurityContext sc) {
+        EventFeedStreamer streamer = new EventFeedStreamer(contest, controller, servletRequest, sc);
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        streamer.addStream(stream, new EventFeedFilter());
```

Comment:
It would be nice if there was a comment noting that `addStream()` _also outputs all past events to the stream_ (just to help the reader...)  Otherwise, people (like me) will spend time trying to figure out why there are two consecutive statements
```java
streamer.addStream(stream, new EventFeedFilter());
streamer.removeStream(stream);
```
Without knowing that side-effect of `addStream()`, that sequence makes little sense...

(And yes, I'm aware that this is legacy code...)

Reply:
Agreed.  Fixed in next push.